homeassistant/components/velbus/services.py

```diff
@@ -0,0 +1,138 @@
+"""Support for Velbus devices."""
+
+from __future__ import annotations
+
+from contextlib import suppress
+import logging
+import os
+import shutil
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_ADDRESS, Platform
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.storage import STORAGE_DIR
+
+from .const import (
+    CONF_INTERFACE,
+    CONF_MEMO_TEXT,
+    DOMAIN,
+    SERVICE_CLEAR_CACHE,
+    SERVICE_SCAN,
+    SERVICE_SET_MEMO_TEXT,
+    SERVICE_SYNC,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [
+    Platform.BINARY_SENSOR,
+    Platform.BUTTON,
+    Platform.CLIMATE,
+    Platform.COVER,
+    Platform.LIGHT,
+    Platform.SELECT,
+    Platform.SENSOR,
+    Platform.SWITCH,
+]
+
+
+def cleanup_services(hass: HomeAssistant) -> None:
+    """Unregister the velbus services."""
+    hass.services.async_remove(DOMAIN, SERVICE_SCAN)
+    hass.services.async_remove(DOMAIN, SERVICE_SYNC)
+    hass.services.async_remove(DOMAIN, SERVICE_SET_MEMO_TEXT)
+    hass.services.async_remove(DOMAIN, SERVICE_CLEAR_CACHE)
+
+
+def setup_services(hass: HomeAssistant) -> None:
+    """Register the velbus services."""
+
+    def check_entry_id(interface: str) -> str:
+        for config_entry in hass.config_entries.async_entries(DOMAIN):
+            if "port" in config_entry.data and config_entry.data["port"] == interface:
+                return config_entry.entry_id
+        raise vol.Invalid(
+            "The interface provided is not defined as a port in a Velbus integration"
+        )
```

Comment:
Ideally we don't have this as a transformer in the data validation, but rather in the execution, to also catch not loaded entries or not found entries

Reply:
Yes i know, i'm working on some new services for velbus, in the PR we will migrate away from this old system